### PR TITLE
Add high-level documentation about Java traces [DOC-128]

### PR DESCRIPTION
**Pull request by jeannielynnmoulton**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- ~[] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).~ not applicable
- ~[] My commit message includes data points confirming performance improvements (if claimed).~ not applicable
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 0e696e842f8921145b1d26d766aeaa569163cd8a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91254685
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e5d9bd97d9c6e6e156b5aaadfce9f820461e88f6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91255414

File: src/goto-programs/README.md [hunk 1]
```diff
@@ -509,3 +509,80 @@ C++ modules:
 Dependencies:
   - [linking folder](\ref linking).
   - [typecheck](\ref section-goto-typecheck).
+
+\section java-trace-structure Java Trace Structure
```
Comment (tautschnig) [inline src/goto-programs/README.md]: I'm not sure I agree that talking about Java should happen in this folder. Looking at the contents, however, it doesn't particularly Java-specific anyway. Can the title and the introduction below be adjusted a bit to make clear that it's only about the example being in Java, which seems perfectly fine to me?

Comment (peterschrammel) [inline src/goto-programs/README.md]: Shouldn't this be called 'Goto Trace Structure' ?
A goto trace corresponds to a goto program in the first place; the link to C or Java is an interpretation of it... So, I'd start with the generic description and then add separate sections to detail language-specific interpretations of it.

File: src/goto-programs/README.md [hunk 2]
```diff
@@ -509,3 +509,80 @@ C++ modules:
...
+* \ref goto_trace_stept
+* \ref goto_trace_stept::typet
+
+A trace represents the execution of a program as a series of steps. The class 
+\ref goto_tracet contains an ordered list of \ref goto_trace_stept, each 
+representing a step. The main step types used in a Java program are:
+- function call
+- function return
+- assignment
+
```
Comment (peterschrammel) [inline src/goto-programs/README.md:553]: assert?

File: src/goto-programs/README.md [hunk 3]
```diff
@@ -509,3 +509,80 @@ C++ modules:
...
+An assignment step has the goto_trace_stept::typet 
+\ref goto_trace_stept::typet::ASSIGNMENT and signifies that the program is 
+making an assignment to a variable. For an assignment LHS = RHS, the step 
+contains the symbol representing the LHS
+(\ref goto_trace_stept::full_lhs) and the value expression representing the RHS 
+(\ref goto_trace_stept::full_lhs_value).
+
+The expressions on each side vary depending on the Java type, whether its a
+member assignment, whether its an assignment by reference, etc. Some examples:
+- A global variable assignment will have a LHS \ref symbolt containing the
```
Comment (peterschrammel) [inline src/goto-programs/README.md]: The Java specific section should start here.

File: src/goto-programs/README.md [hunk 4]
```diff
@@ -509,3 +509,80 @@ C++ modules:
...
+\subsection assignment-step-structure Assignment
+
+An assignment step has the goto_trace_stept::typet 
+\ref goto_trace_stept::typet::ASSIGNMENT and signifies that the program is 
+making an assignment to a variable. For an assignment LHS = RHS, the step 
+contains the symbol representing the LHS
+(\ref goto_trace_stept::full_lhs) and the value expression representing the RHS 
+(\ref goto_trace_stept::full_lhs_value).
+
+The expressions on each side vary depending on the Java type, whether its a
```
Comment (peterschrammel) [inline src/goto-programs/README.md]: The same is true for plain GOTO (which has structs, arrays, etc).

File: src/goto-programs/README.md [hunk 5]
```diff
@@ -509,3 +509,80 @@ C++ modules:
...
+the containing class which contains an identifier for the containing class, a
+RHS expression containing the value. Inherited members have a nested member
+structure. Members of members are assigned using a series of single member
+assignments.
+
+\subsection other-steps Other steps
+
+In a trace of a Java program, there are other step types that are internal
+including:
+- goto_trace_stept::typet::ASSERT (failure), used in bytecode instrumentation
```
Comment (peterschrammel) [inline src/goto-programs/README.md]: ...and properties

**Review by peterschrammel — COMMENTED** [review]
I'd like to see this explained wrt 'plain' GOTO first. The Java-specific things should be in a separate section or preferably even in JBMC, e.g. jbmc/src/java_bytecode/README.md

File: src/goto-programs/README.md [hunk 6]
```diff
@@ -509,3 +509,80 @@ C++ modules:
...
+\ref goto_trace_stept::typet, but this overview will focus on the main types 
+used in Java.
+
+\subsection general-step-structure General Step Structure
+
+Every goto_trace_stept has a type (\ref goto_trace_stept::typet), e.g. function
+call. The type of a step can be checked using the boolean functions such as 
+\ref goto_trace_stept::is_function_call(). 
+
+The remainder of the members of goto_tracet depend on the step type.
```
Comment (smowton) [inline src/goto-programs/README.md]: The meaning of the remainder of the members of `goto_tracet_stept`?

File: src/goto-programs/README.md [hunk 7]
```diff
@@ -509,3 +509,80 @@ C++ modules:
...
+\ref goto_trace_stept::typet::ASSIGNMENT and signifies that the program is 
+making an assignment to a variable. For an assignment LHS = RHS, the step 
+contains the symbol representing the LHS
+(\ref goto_trace_stept::full_lhs) and the value expression representing the RHS 
+(\ref goto_trace_stept::full_lhs_value).
+
+The expressions on each side vary depending on the Java type, whether its a
+member assignment, whether its an assignment by reference, etc. Some examples:
+- A global variable assignment will have a LHS \ref symbolt containing the
+Java type information and an identifier that does not include a function
```
Comment (smowton) [inline src/goto-programs/README.md]: Perhaps just an identifier corresponding to a symbol with `is_static_lifetime` set?

File: src/goto-programs/README.md [hunk 8]
```diff
@@ -509,3 +509,80 @@ C++ modules:
...
+The remainder of the members of goto_tracet depend on the step type.
+
+\subsection function-call-step-structure Function Calls
+
+A function call step has the goto_trace_stept::typet 
+\ref goto_trace_stept::typet::FUNCTION_CALL 
+and signifies that the program is calling a function.
+The step contains a valid identifier of the function being called 
+(\ref goto_trace_stept::called_function) and a vector of the function arguments 
+(\ref goto_trace_stept::function_arguments).
```
Comment (thk123) [inline src/goto-programs/README.md]: :pick: Mention that `\ref goto_trace_stept::function` is the function of the call site

File: src/goto-programs/README.md [hunk 9]
```diff
@@ -509,3 +509,80 @@ C++ modules:
...
+
+\subsection function-return-step-structure Function Return
+
+A function return step has the goto_trace_stept::typet 
+\ref goto_trace_stept::typet::FUNCTION_RETURN and signifies that the program is 
+exiting a function. There is no special information contained in a function 
+return step.
```
Comment (thk123) [inline src/goto-programs/README.md]: :pick: Mention that `\ref goto_trace_stept::function` is the function being returned from.

File: src/goto-programs/README.md (hunk 3 above)
Comment (thk123) [inline src/goto-programs/README.md]: :pick: `\ref symbol_exprt`

**Review by thk123 — APPROVED** [review]
Nothing blocking (though splitting the Java specific stuff out would be good). Perhaps in a follow up doc day we should explain about some of the multi-step structure (like how a function call is followed by assignments to each of the parameters).

File: src/goto-programs/README.md (hunk 2 above)
Comment (jeannielynnmoulton) [reply src/goto-programs/README.md:553]: This is mentioned below.

**Comment by jeannielynnmoulton** [comment]
@peterschrammel I've addressed your comments, can you please re-review (there's a separate commit for convenience)? @thk123 I am working on adding some docs for more complex objects but I would like to do those in another PR and perhaps in a different location to keep things moving.

File: src/goto-programs/README.md [hunk 10]
```diff
@@ -534,3 +534,101 @@ C++ modules:
...
+of the call site (\ref goto_trace_stept::function).
+
+Following a function call, \ref assignment-step-structure assignments are made
+to each parameter, if there are any, including the `this` parameter.
+
+\subsection function-return-step-structure Function Return
+
+A function return step has the goto_trace_stept::typet 
+\ref goto_trace_stept::typet::FUNCTION_RETURN and signifies that the program is 
+exiting a function. \ref goto_trace_stept::function is the function being
```
Comment (thk123) [inline src/goto-programs/README.md]: This becomes `goto_trace_stept::function_id` as doxygen error has pointed out.

**Review by peterschrammel — APPROVED** [review]
CI is still failing.
Also, please squash commits.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: bcee5b328bad78a7df0512c43886ae71c601fc21).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99838363
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4dc4b65441f3026f2a3530317da3afd5826bb711).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99838708

**Pull request merged**